main.nf

```diff
@@ -0,0 +1,533 @@
+
+    input:
+    file(fasta) from fastaCh3
+    file(fastagz) from tmpFastaGzCh
+
+    output:
+    file("*.gz.fai") into gzFaiCh
+
+    script:
+    """
+    samtools faidx $fastagz
+    """
+  }
+}
+
+if(!gzi){
+  process preprocessGZI {
+    publishDir "$baseDir/sampleDerivatives"
+
+    input:
+    file(fasta) from fastaCh4
+
+    output:
+    file("*.gz.gzi") into gziCh
+
+    script:
+    """
+    bgzip -c -i ${fasta} > ${fasta}.gz
+    """
+  }
+}
+
+//channel to collect all necessary reference files
+fastaChannel = Channel.from(fastaCh).mix(bedCh, faiCh, fastaGzCh, gzFaiCh, gziCh).collect()
+
+/********************************************************************
+  process preprocessBAM
+  Takes care of the read group line.
+********************************************************************/
+
+process preprocessBAM{
+
+  tag "${bam[0]}"
+  publishDir "$baseDir/sampleDerivatives"
+
+  input:
+  set val(prefix), file(bam) from bamChannel
+
+  output:
+  set file("ready/${bam[0]}"), file("ready/${bam[0]}.bai") into completeChannel
+
+  script:
+  """
+  mkdir ready
+  [[ `samtools view -H ${bam[0]} | grep '@RG' | wc -l`   > 0 ]] && { mv $bam ready;}|| { picard AddOrReplaceReadGroups \
+  I=${bam[0]} \
+  O=ready/${bam[0]} \
+  RGID=${params.rgid} \
+  RGLB=${params.rglb} \
+  RGPL=${params.rgpl} \
+  RGPU=${params.rgpu} \
+  RGSM=${params.rgsm};}
+  cd ready ;samtools index ${bam[0]};
+  """
+}
+
+
+//make tupule of reference files using fasta file as identifier
+fastaChannel.map{file -> tuple (1,file[0],file[1],file[2],file[3],file[4], file[5])}
+            .set{all_fa};
+
+completeChannel.map { file -> tuple(1,file[0],file[1]) }
+               .set{all_bam};
+
+all_fa.cross(all_bam)
+      .set{all_fa_bam};
+
+      /********************************************************************
+        process makeExamples
+        Getting bam files and converting them to images ( named examples )
+
+	Can be parallelized through the params.n_shards
+	( if params.n_shards >= 1 parallelization happens automatically)
+      ********************************************************************/
+
+  process makeExamples{
+
+    tag "${bam[1]}"
+
+    input:
+    set file(fasta), file(bam) from all_fa_bam
+
+    output:
+    set file("${fasta[1]}"),file("${fasta[1]}.fai"),file("${fasta[1]}.gz"),file("${fasta[1]}.gz.fai"), file("${fasta[1]}.gz.gzi"),val("${bam[1]}"), file("shardedExamples") into examples
+
+    script:
+    """
+    mkdir logs
+    mkdir shardedExamples
+    dv_make_examples.py \
+    --cores ${task.cpus} \
+    --sample ${bam[1]} \
+    --ref ${fasta[1]}.gz \
+    --reads ${bam[1]} \
+    --regions ${fasta[2]} \
+    --logdir logs \
+    --examples shardedExamples
+    """
+  }
+/********************************************************************
+  process call_variants
+  Doing the variant calling based on the ML trained model.
+********************************************************************/
+
+process call_variants{
```

Comment:
Just a personal note:
You were using `mixedCase` for naming your processes, and now you switched to `snake_case`
I know it's not important at all, and I feel sorry for mentioning it, but I do find that disturbing...
Can you please stick with one method?